### PR TITLE
Implmentation of dense_sst_to_dst and sst_dst_to_dense

### PR DESCRIPTION
**Pull request by riohib**
## What does this PR do?
1. Implements the dense_sst_to_dst and sst_dst_to_dense method.
2. Adds tests for perfect reconstruction with all top-k across different dims.
3. Adds tests for the two new methods.

## Before submitting

- [x] Did you have fun?
  - Make sure you had fun coding 🙃
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [x] N/A
- [ ] Did you make sure to update the docs?
  - [x] N/A
- [x] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [x] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 1]
```diff
@@ -157,6 +170,7 @@ def __init__(
         self._validate_conf()
         # TODO (Min): Type checking for the following
         self._transform = torch.fft.fft if algo is Algo.FFT else _dct_transform  # type: ignore
+        self._inverse_transform = torch.fft.ifft if algo is Algo.FFT else _inverse_dct_transform  # type: ignore
```
Comment (riohib) [inline fairscale/experimental/wgit/signal_sparsity.py:173]: Would condensing this in a single `if algo is Algo.FFT: ...` be better, or is this fine?

Comment (min-xu-ai) [reply fairscale/experimental/wgit/signal_sparsity.py:173]: this is fine. you could do

a, b = fft, ifft if algo is fft else dct, idct

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 2]
```diff
@@ -88,6 +88,19 @@ def _dct_transform(dense: Tensor) -> Tensor:
     raise NotImplementedError("Support for DCT has not been implemented yet!")
 
 
+def _inverse_dct_transform(dense: Tensor) -> Tensor:
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:91]: ### 
```suggestion
def _inverse_dct_transform(sst: Tensor) -> Tensor:
```

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 3]
```diff
@@ -88,6 +88,19 @@ def _dct_transform(dense: Tensor) -> Tensor:
...
+    """Should take a tensor and perform an inverse Discrete Cosine Transform on the tensor.
+
+    Args:
+        dense (Tensor):
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:95]: ### 
```suggestion
        sst (Tensor):
```

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 4]
```diff
@@ -88,6 +88,19 @@ def _dct_transform(dense: Tensor) -> Tensor:
     raise NotImplementedError("Support for DCT has not been implemented yet!")
 
 
+def _inverse_dct_transform(dense: Tensor) -> Tensor:
+    """Should take a tensor and perform an inverse Discrete Cosine Transform on the tensor.
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:92]:
```suggestion
    """Should take a tensor and perform an inverse Discrete Cosine Transform and return a new tensor.
```

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 5]
```diff
@@ -88,6 +88,19 @@ def _dct_transform(dense: Tensor) -> Tensor:
...
+
+    Args:
+        dense (Tensor):
+            Input dense tensor (no zeros).
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:96]:
```suggestion
            Input sst tensor (may have zeros) in frequency domain.
```

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 6]
```diff
@@ -88,6 +88,19 @@ def _dct_transform(dense: Tensor) -> Tensor:
...
+    Returns:
+        (Tensor):
+            transformed dense tensor with iDCT values
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:99]:
```suggestion
            A new, transformed dense tensor with real domain values.
```

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 7]
```diff
@@ -250,32 +261,29 @@ def dense_sst_to_dst(self, dense: Tensor, sst: Tensor) -> Tensor:
             (Tensor):
                 Same shaped tensor, still dense format but has zeros. Non-zeros are top-k delta values.
         """
-        pass
+        delta = dense - self.sst_dst_to_dense(sst)  # sst_dst_to_dense(sst) returns the inverse transform here
+        top_k_total_size = _top_k_total_size(dense, self._dst_top_k_dim)
+        k = _get_k_for_topk(self._dst_top_k_percent, self._dst_top_k_element, top_k_total_size)
+        return _scatter_topk_to_sparse_tensor(delta.abs(), delta, k, dim=self._dst_top_k_dim)
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:267]: super nice that we can reuse this function

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 8]
```diff
@@ -250,32 +261,29 @@ def dense_sst_to_dst(self, dense: Tensor, sst: Tensor) -> Tensor:
             (Tensor):
                 Same shaped tensor, still dense format but has zeros. Non-zeros are top-k delta values.
         """
-        pass
+        delta = dense - self.sst_dst_to_dense(sst)  # sst_dst_to_dense(sst) returns the inverse transform here
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:264]: I would move this line to after 266 just before its use.
Also, I would add:

del dense

after this line so that the dense tensor could be freed it no other reference to it exists to save memory.

File: fairscale/experimental/wgit/signal_sparsity.py [hunk 9]
```diff
@@ -250,32 +261,29 @@ def dense_sst_to_dst(self, dense: Tensor, sst: Tensor) -> Tensor:
...
             (Tensor):
                 A dense tensor in real number domain from the SST.
         """
-        pass
-
-    def sst_or_dst_to_mask(self) -> None:
-        # we shouldn't need this function since going from SST/DST to mask should be a
-        # trivial call in pytorch. Maybe I am missing something.
-        pass
+        dense_recons = self._inverse_transform(sst).real
```
Comment (min-xu-ai) [inline fairscale/experimental/wgit/signal_sparsity.py:283]: do FFT and DCT both have .real after inverse? 

Also, should inverse function always return the real part since the complex part should be zero, right?

If so, just absorb the `.real` part into the inverse transform function and caller will be simpler.

**Review by min-xu-ai — COMMENTED** [review]
Very nice. Will review the test code next.

For the API functions, please do some tensor shape check first and raise ValueError is shapes are wrong.

File: tests/experimental/wgit/test_signal_sparsity.py [hunk 10]
```diff
@@ -147,3 +191,42 @@ def test_percent_element(tensor, expected, dim, percent, k):
...
+        (torch.arange(12).float().reshape(4, 3), 12, None),  # top-12, dim=None
+        (torch.arange(12).float().reshape(4, 3), 4, 0),  # top-4, dim=0
+        (torch.arange(12).float().reshape(4, 3), 3, 1),  # top-3, dim=1
+        (torch.arange(60).float().reshape(10, 6), 60, None),  # top-60, dim=None
+        (torch.arange(60).float().reshape(10, 6), 10, 0),  # top-10, dim=0
+        (torch.arange(60).float().reshape(10, 6), 6, 1),  # top-6, dim=1
+        (torch.arange(60).float().reshape(2, 5, 6), 5, 1),  # top-5, dim=1
+    ],
+)
+def test_sst_dst_to_dense_reconstruction(dense, k, dim):
```
[the remark this answers is not in the document]
Comment (min-xu-ai) [reply tests/experimental/wgit/test_signal_sparsity.py:216]: nice test by the way

File: fairscale/experimental/wgit/signal_sparsity.py (hunk 7 above)
Comment (riohib) [reply fairscale/experimental/wgit/signal_sparsity.py:267]: Yeah, I found this use of the function to be quite nice as well!

File: fairscale/experimental/wgit/signal_sparsity.py (hunk 9 above)
Comment (riohib) [reply fairscale/experimental/wgit/signal_sparsity.py:283]: with the DCT, we would either have to implement it with native pytorch operations, or use scipy. I don't think it would have a `.real` attribute by default, or we wanted to implement DCT as a class. Either way, seems like absorbing `real` would be better. And DCT is always real if I am not wrong.

File: tests/experimental/wgit/test_signal_sparsity.py [hunk 11]
```diff
@@ -29,7 +30,15 @@ def get_test_params():
         dtype=torch.complex64,
     )
 
-    expected_4x3_0 = torch.tensor(
+    expd_dst_4x3_None = torch.tensor(
+        [[0.0, 0.0, 0.0], [0.0, 4.0, 5.0], [0.0, 0.0, 0.0], [0.0, 0.0, 0.0]], dtype=torch.float32
+    )  # with dim=None, top-2
+
+    expd_wr_4x3_None = torch.tensor(
```
[the remark this answers is not in the document]
Comment (riohib) [reply tests/experimental/wgit/test_signal_sparsity.py:37]: RT is an appropriate name!

[the remark this answers is not in the document]
Comment (riohib) [reply tests/experimental/wgit/test_signal_sparsity.py:37]: Also, skipped dim and top info on RT tensors because we are not doing any top-k operations on them or to get them.

[the remark this answers is not in the document]
Comment (riohib) [reply tests/experimental/wgit/test_signal_sparsity.py:37]: Instead added sst and dst topk and dim values used to get the RT

File: fairscale/experimental/wgit/signal_sparsity.py (hunk 9 above)
Comment (riohib) [reply fairscale/experimental/wgit/signal_sparsity.py:283]: I think we can simply do `dense_recons = torch.real(self._inverse_transform(sst))` which will tackle both cases

**Review by min-xu-ai — APPROVED** [review]
awesome
